### DISCUSSION TITLE
multi-channel with roc-vad?

### DISCUSSION THREAD

**Discussion by paulharter** [roc-streaming/roc-toolkit#785] [Q&A]
Hi,

I see that the roc-toolkit supports multi-channel. I would like to be able to create multi-channel virtual mac devices with roc-vad, but it currently only seems to support mono or stereo.

Is this supported but undocumented?
Is this theoretically possible but just not implemented yet?
Is it just not possible as libASPL doesn't support it?
Is this not possible because CoreAudio plugins don't support it?

If it is possible I could have a look at making it work, any suggestion and pointers about where to start would be a great help, but if it's a known problem that can't be done it would be nice to know that too.

Thanks,

Paul

**Comment by gavv** [comment]
Hi,

Short answer: not supported, but possible.

libASPL should support arbitrary number of channels, though I think I never tested that.

roc-toolkit C API allows to select MULTITRACK layout and then set `tracks` to the desired number of channels. See here: https://roc-streaming.org/toolkit/docs/api/reference.html#c.roc_channel_layout.ROC_CHANNEL_LAYOUT_MULTITRACK

> If it is possible I could have a look at making it work, any suggestion and pointers about where to start would be a great help, but if it's a known problem that can't be done it would be nice to know that too.

You're welcome!

Here are the steps comes to mind:

1. Add multi-channel support to driver_protocol.proto. Since roc-vad rpc mostly mirrors C API, we can add multitrack layout to RvChannelLayout, and track_count field to RvDeviceEncoding and RvPacketEncoding, similar to C API.

2. Update rpc_serdes.cpp to read/write track count

3. Update compute_channel_count() in device.cpp

4. Update CLI: cmd_device_add_receiver.cpp, cmd_device_add_sender.cpp, print.cpp

Note: to employ multi-channel mode, user will need to register custom encoding on both sender and receiver, see here: https://github.com/roc-streaming/roc-vad?tab=readme-ov-file#custom-packet-encoding

**Comment by paulharter** [comment]
@gavv - That's good news. Thank you for the detailed reply with what would need doing. I'll have a go at it next week.

**Reply by gavv** [reply]
Awesome and you're welcome

**Comment by paulharter** [comment]
Hey @gavv  - I've done a pull request for this.

I'm afraid my C++ is a bit rusty so there might be ugly things in there, but it seems to work ok. I mostly did exactly as you suggested but instead of changing compute_channel_count() I pre populated the channel_count when parsing the rpc in the case of multitrack., this avoided extending DeviceLocalEncoding .

Thanks for your help with this,

P

**Reply by gavv** [reply]
The patch is good, merged!
